crates/engine/src/response/data.rs

```diff
@@ -289,6 +319,40 @@ impl DataPart {
             map_id,
         }
     }
+
+    pub fn push_string(&mut self, s: String) -> PartString {
+        let len = s.len() as u32;
+        let ptr = PartStrPtr(s.as_ptr());
+        self.strings.push(s);
+        let out = PartString {
+            part_id: self.id,
+            ptr,
+            len,
+        };
+        debug_assert!(self.deref_part_string(out) == self.strings[self.strings.len() - 1]);
+        out
+    }
+
+    /// # Safety
+    /// Must only be used on borrowed data from the deserialized bytes.
+    pub unsafe fn push_borrowed_str(&mut self, s: &str) -> PartString {
+        let len = s.len() as u32;
+        let ptr = PartStrPtr(s.as_ptr());
+        let out = PartString {
+            part_id: self.id,
+            ptr,
+            len,
+        };
+        debug_assert!(self.deref_part_string(out) == s);
+        out
+    }
+
+    fn deref_part_string(&self, PartString { part_id, ptr, len }: PartString) -> &str {
```

Comment:
Could this be:
```suggestion
    fn deref_part_string<'a>(&self, &'a PartString { part_id, ptr, len }: PartString) -> &'a str {
```
?

Reply:
Could you wrap the str lifetime to the DataPart somehow? That would make this a bit safer. Or then add a miri run to the CI.